pkg/lib/bundlev2/synthesis.go

```diff
@@ -0,0 +1,375 @@
+				return nil
+			}
+
+			if !IsSupported(bufObj.GetKind()) {
+
+				if bufObj.GetKind() == CSVKind {
+					b.existingCSVPath = path
+					return b.appendTostructuredObjectMap(CSVKind, bufObj, v1alpha1.ClusterServiceVersion{}, info.Name())
+				}
+
+				if bufObj.GetKind() == CRDKind {
+					return b.appendTostructuredObjectMap(CRDKind, bufObj, v1alpha1.CustomResourceDefinitions{},
+						info.Name())
+				}
+
+				if bufObj.GetKind() == ClusterRoleKind {
+					return b.appendTostructuredObjectMap(ClusterRoleKind, bufObj, rv1.ClusterRole{}, info.Name())
+				}
+
+				if bufObj.GetKind() == ClusterRoleBindingKind {
+					return b.appendTostructuredObjectMap(ClusterRoleBindingKind, bufObj, rv1.ClusterRoleBinding{},
+						info.Name())
+				}
+
+				if bufObj.GetKind() == ServiceKind {
+					return b.appendTostructuredObjectMap(ServiceKind, bufObj, cv1.Service{}, info.Name())
+				}
+
+				if bufObj.GetKind() == ServiceAccountKind {
+					return b.appendTostructuredObjectMap(ServiceAccountKind, bufObj, cv1.ServiceAccount{}, info.Name())
+				}
+
+				if bufObj.GetKind() == RoleKind {
+					return b.appendTostructuredObjectMap(RoleKind, bufObj, rv1.Role{}, info.Name())
+				}
+
+				if bufObj.GetKind() == RoleBindingKind {
+					return b.appendTostructuredObjectMap(RoleBindingKind, bufObj, rv1.RoleBinding{}, info.Name())
+				}
+
+				if bufObj.GetKind() == RegistryV2CSVKind {
+					return b.appendTostructuredObjectMap(RegistryV2CSVKind, bufObj, RegistryV2CSV{}, info.Name())
+				}
+
+				if bufObj.GetKind() == DeploymentKind {
+					return b.appendTostructuredObjectMap(DeploymentKind, bufObj, appsv1.Deployment{}, info.Name())
+				}
+
+				// Skipping kinds of resources (PrometheusRule and ServiceMonitor) not needed for the CSV.
+				return nil
+			}
+			b.logger.Warnf("file %s with kind %s is not supported", info.Name(), bufObj.GetKind())
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+// appendTostructuredObjectMap converts the unstructured object and try to convert it into a specified interface
+// before appending to the structuredObjectMap. It only returns error if the unstructured is not convertible.
+func (b *bundlev2) appendTostructuredObjectMap(kind string, bufObj unstructured.Unstructured, obj interface{},
+	objSource string) error {
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(bufObj.UnstructuredContent(), &obj); err != nil {
+		return fmt.Errorf("error decoding %s as %s kind, %v", objSource, bufObj.GetKind(), err)
+	}
+
+	if obj, ok := b.structuredObjectMap[kind]; !ok {
+		b.structuredObjectMap[kind] = &[]interface{}{obj}
+	} else {
+		*obj = append(*obj, bufObj)
+	}
+
+	return nil
+}
+
+// getRegistryV2CSV gets the RegistryV2CSV object from appendTostructuredObjectMap and ensure that there is only one of
+// its kind.
+func (b *bundlev2) getRegistryV2CSV() (*RegistryV2CSV, error) {
+	if csvBasicsIF, ok := b.structuredObjectMap[RegistryV2CSVKind]; ok {
+		if len(*csvBasicsIF) != 1 {
```

Comment:
Why are we adding an obj to a slice of values in the map, and then erroring later if the length of the slice is greater than 1? Wouldn't simply having a one -> one key/value map be more appropriate?

Reply:
yeah, but this map contains other objects that could have more than one item. The singular objects are, **for now**, deployment and noninferrableCSV.

Reply:
I guess, by the end, if those 2 are the only singletons, I'll make the change.

Reply:
I'll change it